File: 239-sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int [] result = new int[nums.length-k+1];
        PriorityQueue<Pair> pq = new PriorityQueue<>();
        for(int i = 0;i<k;i++){
            pq.add(new Pair(nums[i],i));
        }
        int left = 0;
        int right = k-1;
        while(right < nums.length-1){
            while(!(pq.peek().index >= left && pq.peek().index <= right)){
                pq.poll();
            }
            result[left] = pq.peek().value;
            left++;
            right++;
            pq.add(new Pair(nums[right],right));
        }
        while(!(pq.peek().index >= left && pq.peek().index <= right)){
            pq.poll();
        }
        result[left] = pq.peek().value;
        return result;
    }
}
class Pair implements Comparable<Pair>{
    int value;
    int index;
    public Pair(int val, int index){
        this.value = val;
        this.index = index;
    }
    public int compareTo(Pair other){
        return other.value - this.value;
    }
}

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = deque()
        result = []
        right = k-1
        left = 0
        for i in range(k):
            while q and q[-1] < nums[i]:
                q.pop()
            q.append(nums[i])
        while right < len(nums):
            # print(q)
            result.append(q[0])
            if q[0] == nums[left]:
                q.popleft()
            left+=1
            right+=1
            if right < len(nums):
                while q and q[-1] < nums[right]:
                    q.pop()
                q.append(nums[right])
        return result